### PR TITLE
Refactoring to make cardinality estimation provider-specific

### PR DESCRIPTION
**Pull request by alephonea**
### Changelog entry 

...

### Changelog category 

* New feature
* Experimental feature
* Improvement
* Performance improvement
* Bugfix 
* Backward incompatible change
* Documentation (changelog entry is not required)
* Not for changelog (changelog entry is not required)

### Additional information

...

### CODE REVIEW

**Review by pavelvelikhov — CHANGES_REQUESTED** [review]

File: ydb/library/yql/core/cbo/cbo_optimizer_new.cpp [hunk 1]
```diff
@@ -109,4 +109,123 @@ void TJoinOptimizerNode::Print(std::stringstream& stream, int ntabs) {
...
+
+    for(size_t i=0; i<stats.KeyColumns.size(); i++){
+        if (std::find(joinKeys.begin(), joinKeys.end(), stats.KeyColumns[i]) == joinKeys.end()) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool TDummyProviderContext::IsJoinApplicable(const std::shared_ptr<IBaseOptimizerNode>& left,
```
Comment (pavelvelikhov) [inline ydb/library/yql/core/cbo/cbo_optimizer_new.cpp:125]: Lets rename TDummyProviderContext to something like TBaseProviderContext or TDefaultProviderContext.

Its not that dummy anymore if we're including the standard cardinality estimation in there

**Review by pavelvelikhov — CHANGES_REQUESTED** [review]
We also need to fix our KqpProviderContext to subclass from TDummyProviderContext (renamed to something less insulting)

**Review by pavelvelikhov — APPROVED** [review]

File: ydb/library/yql/core/cbo/cbo_optimizer_new.cpp (hunk 1 above)
Comment (alephonea) [reply ydb/library/yql/core/cbo/cbo_optimizer_new.cpp]: Done.

**Comment by alephonea** [comment]
> We also need to fix our KqpProviderContext to subclass from TDummyProviderContext (renamed to something less insulting)

**Pull request closed without merging**